libs/contrib/Data/Nat/Factor.idr

```diff
@@ -0,0 +1,333 @@
+module Data.Nat.Factor
+
+import Data.Fin
+import Data.Fin.Extra
+import Data.Nat
+import Syntax.PreorderReasoning
+
+%default total
+%access public export
+
+
+||| Factor n p is a witness that p is indeed a factor of n,
+||| i.e. there exists a q such that p * q = n.
+data Factor : Nat -> Nat -> Type where
+    CofactorExists : {n, p : Nat} -> (q : Nat) -> LTE 1 n -> p * q = n -> Factor n p
```

Comment:
I notice that there are a lot of places with `rewrite sym prf` or
something like that. I think the `sym`s are requried because this
proof has the form `p * q = n`. If it were `n = p * q`, I think the
`sym`s could be done away with, which would be a little less verbose.

Generally proof equations are written as complexity-reducing devices,
with the more complex part on the left and the less complex on the
right. For example, the monoid neutral law looks like `n + 0 = n`,
because `n` is simpler than `n + 0`. But with factors, the reduction
in complexity is to say that a larger number can be written as the
product of smaller numbers, so I think `n = p * q` is appropriate.

(It might be too late to implement this minor suggestion.)

Reply:
That's a fair point. And I don't think it's too late for the change. It's only 300 lines of code. I have done much larger refactorings with much less type-checking to support me, so I believe I will do it.